scene/resources/particles_material.cpp

```diff
@@ -282,6 +292,77 @@ void ParticlesMaterial::_update_shader() {
+		code += "uniform vec3 turbulence_noise_speed = vec3(1.0, 1.0, 1.0);\n";
+		if (tex_parameters[PARAM_TURB_INFLUENCE_OVER_LIFE].is_valid()) {
+			code += "uniform sampler2D turbulence_influence_over_life;\n";
+		}
+		if (turbulence_color_ramp.is_valid()) {
+			code += "uniform sampler2D turbulence_color_ramp;\n";
+		}
+		code += "\n";
+
+		//functions for 3D noise / turbulence
+		code += "\n\n";
+		code += "// 3D Noise with friendly permission by Inigo Quilez\n";
+		code += "vec3 hash_noise( vec3 p ) {\n";
+		code += "	p *= mat3(vec3(127.1, 311.7, -53.7), vec3(269.5, 183.3, 77.1), vec3(-301.7, 27.3, 215.3));\n";
+		code += "	return 2.0 * fract(fract(p)*4375.55) -1.;\n";
+		code += "}\n";
+		code += "\n";
+		code += "float noise( vec3 p) {\n";
+		code += "	vec3 i = floor(p);;\n";
+		code += "	vec3 f = fract(p);\n ";
+		code += "	vec3 u = f * f * (3.0 - 2.0 * f);\n";
+		code += "\n";
+		code += "	return 2.0*mix( mix( mix( dot( hash_noise( i + vec3(0.0,0.0,0.0) ), f - vec3(0.0,0.0,0.0) ), dot( hash_noise( i + vec3(1.0,0.0,0.0) ), f - vec3(1.0,0.0,0.0) ), u.x),\n";
+		code += "			mix( dot( hash_noise( i + vec3(0.0,1.0,0.0) ), f - vec3(0.0,1.0,0.0) ), dot( hash_noise( i + vec3(1.0,1.0,0.0) ), f - vec3(1.0,1.0,0.0) ), u.x), u.y),\n";
+		code += "		mix( mix( dot( hash_noise( i + vec3(0.0,0.0,1.0) ), f - vec3(0.0,0.0,1.0) ), dot( hash_noise( i + vec3(1.0,0.0,1.0) ), f - vec3(1.0,0.0,1.0) ), u.x),\n";
+		code += "			mix( dot( hash_noise( i + vec3(0.0,1.0,1.0) ), f - vec3(0.0,1.0,1.0) ), dot( hash_noise( i + vec3(1.0,1.0,1.0) ), f - vec3(1.0,1.0,1.0) ), u.x), u.y), u.z);\n";
+		code += "}\n\n";
+		code += "// Curl 3D and noise_3d function with friendly permission by Isaac Cohen\n";
+		code += "vec3 noise_3d(vec3 p) {\n";
+		code += "	float s = noise(p);\n";
+		code += "	float s1 = noise(vec3(p.y - 19.1, p.z + 33.4, p.x + 47.2));\n";
+		code += "   float s2 = noise(vec3(p.z + 74.2, p.x - 124.5, p.y + 99.4));\n";
+		code += "	vec3 c = vec3(s, s1, s2);\n";
+		code += "	return c;\n";
+		code += "}\n\n";
+		code += "vec3 curl_3d(vec3 p, float c) {\n";
+		code += "	float epsilon = 0.001 + c;\n";
+		code += "	vec3 dx = vec3(epsilon, 0.0, 0.0);\n";
+		code += "	vec3 dy = vec3(0.0, epsilon, 0.0);\n";
+		code += "	vec3 dz = vec3(0.0, 0.0, epsilon);\n";
+		code += "	vec3 x0 = noise_3d(p - dx).xyz;\n";
+		code += "	vec3 x1 = noise_3d(p + dx).xyz;\n";
+		code += "	vec3 y0 = noise_3d(p - dy).xyz;\n";
+		code += "	vec3 y1 = noise_3d(p + dy).xyz;\n";
+		code += "	vec3 z0 = noise_3d(p - dz).xyz;\n";
+		code += "	vec3 z1 = noise_3d(p + dz).xyz;\n";
+		code += "	float x = y1.z - y0.z - z1.y + z0.y;\n";
+		code += "	float y = z1.x - z0.x - x1.z + x0.z;\n";
+		code += "	float z = x1.y - x0.y - y1.x + y0.x;\n";
+		code += "	float divisor = 1.0 / (2.0 * epsilon);\n";
+		code += "	return vec3(normalize(vec3(x, y, z) * divisor));\n";
+		code += "}\n";
+		code += "vec3 get_noise_direction(vec3 pos, vec3 et, vec3 tn) {\n";
```

Comment:
The parameter names should be updated to use more than 2 letters i.e. ``et`` -> ``emission_pos`` and ``tn`` -> ``time_noise``

Reply:
fixed in next update.